Kernel/Arch/aarch64/RPi/Timer.cpp

```diff
@@ -33,10 +33,13 @@ Timer::Timer()
 {
 }
 
+static Timer* s_instance;
+
 Timer& Timer::the()
 {
-    static Timer instance;
-    return instance;
+    if (!s_instance)
+        s_instance = new Timer;
```

Comment:
I'm confused by this commit. Classes with virtual methods can be stored statically just fine: https://godbolt.org/z/q769o56W8

What am I missing?

Reply:
First of all the linker starts to complain that `__dso_handle` and `__cxa_atexit` are not defined. I then define these symbols, and when running the code I hit the following panic:
```
[Kernel]: ASSERTION FAILED: not reached
[Kernel]: ./Kernel/Arch/aarch64/init.cpp:78 in int __cxa_atexit(void (*)(void*), void*, void*)
[Kernel]: KERNEL PANIC! :^(
[Kernel]: Aborted
[Kernel]: at ./Kernel/Arch/aarch64/Panic.cpp:23 in void __assertion_failed(const char*, const char*, unsigned int, const char*)
[Kernel]: Kernel + 0x00000000000e99e4  Kernel::__panic(char const*, unsigned int, char const*) [clone .localalias] +0x54
[Kernel]: Kernel + 0x00000000000e9acc  __assertion_failed(char const*, char const*, unsigned int, char const*) [clone .localalias] +0xdc
[Kernel]: Kernel + 0x00000000000e3018  __cxa_atexit +0x28
[Kernel]: Kernel + 0x00000000000df8d4  Kernel::RPi::Timer::the() +0xd4
[Kernel]: Kernel + 0x00000000000e3260  init +0x240
```
Clearly this is somehow calling the destructor of the `Timer` class, so I thought I had to put it on the heap. Is there another way to solve this?

Reply:
You can use `AK::Singleton<Timer>` to prevent the destructor from being called.

Reply:
Ah great, yeah that works indeed! Thanks for the suggestion, changed to use `AK::Singleton` instead. See 1441b9f325dd627c76e310a9bf598637eac933ef

Reply:
It sounds like exit-time destructors don't work, then? I still think

```
 which in turn makes
the class virtual, which means we have to put it into a AK::Singleton,
which makes sure that the destructor is not called.
```

isn't quite right – this works around a bug where exit-time dtors don't work, for not-yet-understood reasons. Classes with virtual methods normally work just fine in function-local statics, or in locals.

But code is fine as-is. Thanks for explaining why it's needed.